mempool/clist_mempool.go

```diff
@@ -818,26 +882,73 @@ func (mem *CListMempool) NewIterator() Iterator {
 // Unsafe for concurrent use by multiple goroutines.
 func (iter *CListIterator) WaitNextCh() <-chan Entry {
 	ch := make(chan Entry)
-	// Spawn goroutine that waits for the next entry, saves it locally, and puts it in the channel.
 	go func() {
-		if iter.cursor == nil {
-			// We are at the beginning of the iteration or the saved entry got removed: wait until
-			// the list becomes not empty and select the first entry.
-			<-iter.txs.WaitChan()
-			// Note that Front can return nil.
-			iter.cursor = iter.txs.Front()
-		} else {
-			// Wait for the next entry after the current one.
-			<-iter.cursor.NextWaitChan()
-			// If the current entry is the last one or was removed, Next will return nil.
-			iter.cursor = iter.cursor.Next()
-		}
-		if iter.cursor != nil {
-			ch <- iter.cursor.Value.(Entry)
+		// Add the next entry to the channel if not nil.
+		if entry := iter.Next(); entry != nil {
+			ch <- entry.Value.(Entry)
 		} else {
 			// Unblock the receiver (it will receive nil).
 			close(ch)
 		}
 	}()
 	return ch
 }
+
+// PickLane returns a _valid_ lane on which to iterate, according to the WRR algorithm. A lane is
+// valid if it is not empty or the number of accessed entries in the lane has not yet reached its
+// priority value.
+func (iter *CListIterator) PickLane() types.Lane {
+	// Loop until finding a valid lane.
+	for {
+		// If the current lane is not valid, continue with the next lane with lower priority, in a
+		// round robin fashion.
+		lane := iter.mp.sortedLanes[iter.currentLaneIndex]
+		if iter.mp.lanes[lane].Len() == 0 || iter.counters[lane] >= uint32(lane) {
+			// Reset the counter only when the limit on the lane was reached.
+			if iter.counters[lane] >= uint32(lane) {
+				iter.counters[lane] = 0
+			}
+			iter.currentLaneIndex = (iter.currentLaneIndex + 1) % len(iter.mp.sortedLanes)
+			continue
+		}
+		// TODO: if we detect that a higher-priority lane now has entries, do we preempt access to the current lane?
+		return lane
+	}
+}
+
+// Next implements the classical Weighted Round Robin (WRR) algorithm.
+//
+// In classical WRR, the iterator cycles over the lanes. When a lane is selected, Next returns an
+// entry from the selected lane. On subsequent calls, Next will return the next entries from the
+// same lane until `lane` entries are accessed or the lane is empty, where `lane` is the priority.
+// The next time, Next will select the successive lane with lower priority.
+//
+// TODO: Note that this code does not block waiting for an available entry on a CList or a CElement, as
+// was the case on the original code. Is this the best way to do it?
```

Comment:
We will need to profile to see if we are spinning too much (or add a Debug log to check if that log shows up too often)

Reply:
An update on this... we need to improve this logic for sure as, ATM, this is a busy-wait loop (so the channel is useless).

To validate this, I've run an `e2e` test (`simple.toml`) with the following patch (on top of this PR):
```diff
diff --git a/mempool/clist_mempool.go b/mempool/clist_mempool.go
--- a/mempool/clist_mempool.go
+++ b/mempool/clist_mempool.go
@@ -900,6 +900,7 @@ func (iter *CListIterator) WaitNextCh() <-chan Entry {
 func (iter *CListIterator) PickLane() types.Lane {
 	// Loop until finding a valid lane.
 	for {
+		iter.mp.logger.Info("XXX PickLane, beginning of loop")
 		// If the current lane is not valid, continue with the next lane with lower priority, in a
 		// round robin fashion.
 		lane := iter.mp.sortedLanes[iter.currentLaneIndex]
diff --git a/mempool/reactor.go b/mempool/reactor.go
--- a/mempool/reactor.go
+++ b/mempool/reactor.go
@@ -240,10 +240,12 @@ func (memR *Reactor) broadcastTxRoutine(peer p2p.Peer) {
 			return
 		}
 
+		memR.Logger.Info("XXX before select")
 		select {
 		case entry = <-iter.WaitNextCh():
 			// If the entry we were looking at got garbage collected (removed), try again.
 			if entry == nil {
+				memR.Logger.Info("XXX entry is nil")
 				continue
 			}
 		case <-peer.Quit():
diff --git a/test/e2e/networks/simple.toml b/test/e2e/networks/simple.toml
--- a/test/e2e/networks/simple.toml
+++ b/test/e2e/networks/simple.toml
@@ -1,4 +1,2 @@
 [node.validator00]
 [node.validator01]
-[node.validator02]
-[node.validator03]
```

When I ran the testnet, I could see the follwing line being logged endlessly:

```log
validator01  | I[2024-08-09|19:57:12.532] XXX PickLane, beginning of loop              module=mempool
```

Reply:
I have an idea how we could do it. We can talk about it on Mon (to see if you agree on the approach)